### PR TITLE
[wasi] implement ICU bundling

### LINKED ISSUE #84310: WASI app always tries to load icudt.dat from disk
### Description

A default app created using the `wasiconsole` template from the new `wasi-experimental` workload can only run with disk access. Without disk access, it fails with:

```
Failed to load ./icudt.dat
```

This appears to be the case regardless of setting `<InvariantGlobalization>true</InvariantGlobalization>` in the csproj.

The drawback of this is that it prevents single-file deployment or running in any environment that doesn't allow access to arbitrary files on disk.

### Reproduction Steps

* `dotnet new wasiconsole`
* `dotnet build /p:WasmSingleFileBundle=true`
* `wasmtime bin\Debug\net8.0\wasi-wasm\AppBundle\MyApp.wasm`

### Expected behavior

The `icudt.dat` file:

 * Should not be required at all if `InvariantGlobalization` is true on build
 * Should be bundled into the `.wasm` file otherwise

... so in both cases, a default "hello world" app should be able to run without disk access.

### Actual behavior

`Failed to load ./icudt.dat`

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by pavelsavara**
- creates object out of icudt.dat when `InvariantGlobalization!=true` and `WasmSingleFileBundle==true`
- load the ICU data from the object file when `BUNDLED_ASSEMBLIES` is defined

Fixes https://github.com/dotnet/runtime/issues/84310

### CODE REVIEW

File: src/mono/wasi/build/WasiApp.Native.targets [hunk 1]
```diff
@@ -342,19 +342,28 @@
...
     </GetFileHash>
+    <GetFileHash Files="$(MicrosoftNetCoreAppRuntimePackRidNativeDir)icudt.dat">
+      <Output TaskParameter="Items" ItemName="WasmBundleIcuWithHashes" />
+    </GetFileHash>
     <ItemGroup>
       <WasmBundleAssembliesWithHashes Update="@(WasmBundleAssembliesWithHashes)">
         <DestinationFile>$(_WasmIntermediateOutputPath)%(WasmBundleAssembliesWithHashes.Filename)%(WasmBundleAssembliesWithHashes.Extension).$([System.String]::Copy(%(WasmBundleAssembliesWithHashes.FileHash)).Substring(0, 8)).o</DestinationFile>
       </WasmBundleAssembliesWithHashes>
+      <WasmBundleIcuWithHashes Update="@(WasmBundleIcuWithHashes)">
+        <DestinationFile>$(_WasmIntermediateOutputPath)%(WasmBundleIcuWithHashes.Filename)%(WasmBundleIcuWithHashes.Extension).$([System.String]::Copy(%(WasmBundleIcuWithHashes.FileHash)).Substring(0, 8)).o</DestinationFile>
```
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/mono/wasi/build/WasiApp.Native.targets:365]: 8 chars is unique enough. More chars add bytes to size of the download

File: src/mono/wasi/build/WasiApp.Native.targets [hunk 2]
```diff
@@ -376,6 +385,28 @@
...
+
+    <ItemGroup Condition="'$(InvariantGlobalization)' != 'true'">
+      <_WasiObjectFilesForBundle Include="$(_WasmIcuBundleObjectFile)" />
+      <_WasiObjectFilesForBundle Include="%(WasmBundleIcuWithHashes.DestinationFile)" />
+    </ItemGroup>
+
+    <ItemGroup Condition="'$(InvariantGlobalization)' != 'true'">
+      <WasmBundleFileToDelete Remove="$(_WasmIcuBundleObjectFile)" />
+      <WasmBundleFileToDelete Remove="%(WasmBundleIcuWithHashes.DestinationFile)" />
+      <!-- is there better way how to drop icudt.dat from bin folder ? -->
```
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/mono/wasi/build/WasiApp.Native.targets:406]: `WasmBundleFileToDelete` is used in `<Delete` below line 410.

I don`t think that `WasiAppBuilder` is copying files in bin folder, specifically `icudt.dat` is already there and is coming from the runtime pack "copy local". I don't know what would be idiomatic way how to drop it.

File: src/mono/sample/wasi/Directory.Build.targets [hunk 3]
```diff
@@ -13,7 +13,11 @@
       TODOWASI - - tcplisten localhost:64000 - - env DEBUGGER_FD=4
     -->
     <Exec WorkingDirectory="bin/wasi-wasm/AppBundle"
+          Condition="'$(WasmSingleFileBundle)' == 'true'"
           Command="$(WasmtimeDir)wasmtime$(_ExeExt) --dir . $([System.IO.Path]::ChangeExtension($(_SampleAssembly), '.wasm'))" IgnoreExitCode="true" />
+    <Exec WorkingDirectory="bin/wasi-wasm/AppBundle"
+          Condition="'$(WasmSingleFileBundle)' != 'true'"
```
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/mono/sample/wasi/Directory.Build.targets:19]: > Does `WasmSingleFileBundle` control if we want to embed the icu data? 

Yes. In WasmSingleFileBundle we want single file.

> Wasn't embedding planned as the default option? What is the benefit of setting `WasmSingleFileBundle` to false?

That's what Steve proposes. But that requires WASI-SDK to be installed on customer machine and so I think it should not be default.

File: src/mono/wasi/build/WasiApp.Native.targets (hunk 2 above)
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/mono/wasi/build/WasiApp.Native.targets:406]: @radical  please advise

**Comment by pavelsavara** [comment]
> > creates object out of icudt.dat when InvariantGlobalization!=true
> 
> That's good. What about the case when `InvariantGlobalization == true`? 
If you say `/p:InvariantGlobalization=true` at build time (with SDK) it would eliminate the code which is loading it. It would not bundle it either.
If you only set `DOTNET_SYSTEM_GLOBALIZATION_INVARIANT` env var it will just skip loading it at runtime.

File: src/mono/wasi/build/WasiApp.Native.targets (hunk 2 above)
[the remark this answers is not in the document]
Comment (maraf) [reply src/mono/wasi/build/WasiApp.Native.targets:406]: The trick is `<ReferenceCopyLocalPaths Remove="@(_WasmBuildFilesToRemove)" />`

**Review by lewing — APPROVED** [review]

**Pull request merged**